### ISSUE TITLE
Question: regularize for simulation timestep?

### ISSUE THREAD

**Issue by charleswilmot** [patrick-kidger/diffrax#712]
Hello Patrick,

I am seeking for advice. I simulate a dynamical system with diffrax, and I have defined a loss that depends on some parameters **L(theta)**.

I then optimize wrt. these parameters using **LBFGS and Dopri5**.

I could save a lot of time if I could **avoid regions of the parameter space which result in very small timesteps**.

One option could be to make use of dtmin: diffrax.PIDController(dtmin=large_value).

An other interesting option would be to **regularize the loss with a quantity R that penalizes small timesteps**, ie.

L = L(theta) + lambda R(timesteps)
.

I will experiment with dtmin, but I think that I would prefer the latter option.

Do you think something like this would be possible?

**Comment by johannahaffner** [comment]
If you're saving your evolving state at every step, then you could access `solution.ts`, compute differences between successive steps and e.g. add the negative of the log of the step sizes to your loss (`1e-8` becomes pricey, `1e-1` remains cheap).

I've never experimented with that! Perhaps this is also something you'd like to regularise for more or less stringently in the beginning or at the end of training.

**Comment by charleswilmot** [comment]
Thanks Johanna!

I am surprised that the tensor array `solution.ts` is differentiable wrt. inputs (say `theta` for example). Great!

I will experiment with this. In the meantime we can close this question :D

**Comment by johannahaffner** [comment]
Whoops, it probably is not. I just played around with this and the following code produces a zero gradient: 

```python
import jax
import jax.numpy as jnp
import diffrax as dfx
import equinox as eqx
from jaxtyping import Array

def dy(t, y, k):
    return -k * y

class Model(eqx.Module):
    t0: float
    t1: float
    initial_state: Array
    
    def __call__(self, k):
        return dfx.diffeqsolve(
            dfx.ODETerm(dy),
            dfx.Tsit5(),
            t0, t1, 0.01, initial_state, args=k,
            stepsize_controller=dfx.PIDController(rtol=1e-3, atol=1e-6),
            saveat=dfx.SaveAt(steps=True),
        )

@eqx.filter_grad
def objective(k, model):
    solution = model(k)
    
    timepoints = solution.ts
    mask = jnp.isfinite(timepoints)
    masked_timepoints = jnp.where(mask, timepoints, 0.)
    differences = masked_timepoints[1:] - masked_timepoints[0:-1]
    
    return jnp.sum(differences**3)

model = Model(0, 10, jnp.ones(1))
rate = jnp.array(0.5)

model(rate), objective(rate, model)
```

My apologies. In this case, I actually do not have a good answer for you.

**Comment by patrick-kidger** [comment]
Given an ODE
```
dy/dt = f(y)
```
then one standard trick for this is to define
```
dc/dt = ||df/dy||
```
and add c(T) to the loss. Helps to encourage the ODE to not be too wiggly, i.e. not need lots of steps to resolve.

See section 5.4.1.4 of https://arxiv.org/abs/2202.02435, and there were a few papers in the neural ODE literature that focused on this.

**Comment by charleswilmot** [comment]
Amazing, that's easy to implement and has good chances of working!

Thanks to both of you
